RT/Renderer/Backend/DX12/src/Resource.cpp

```diff
@@ -231,40 +231,68 @@ namespace RT
 		ResourceTransition(command_list, dst, dst_state);
 	}
 
-	void UploadTextureData(ID3D12Resource* dst, size_t row_pitch, size_t row_count, const void* data_ptr)
+	void UploadTextureData(ID3D12Resource* dst, size_t width, size_t height, size_t bytes_per_pixel, void *const *mips, size_t mip_count)
 	{
-		D3D12_PLACED_SUBRESOURCE_FOOTPRINT pitch_footprint;
-		D3D12_RESOURCE_DESC dst_desc = dst->GetDesc();
-		size_t dst_byte_size;
-		g_d3d.device->GetCopyableFootprints(&dst_desc, 0, 1, 0, &pitch_footprint, nullptr, nullptr, &dst_byte_size);
+		size_t mip_width  = width;
+		size_t mip_height = height;
+		for (size_t mip_index = 0; mip_index < mip_count; mip_index++)
+		{
+			D3D12_PLACED_SUBRESOURCE_FOOTPRINT pitch_footprint;
+			D3D12_RESOURCE_DESC dst_desc = dst->GetDesc();
+			UINT   dst_row_count;
+			UINT64 dst_row_byte_size;
+			UINT64 dst_byte_size;
+			g_d3d.device->GetCopyableFootprints(&dst_desc, (UINT)mip_index, 1, 0, &pitch_footprint, &dst_row_count, &dst_row_byte_size, &dst_byte_size);
 
-		RingBufferAllocation ring_buf_alloc = AllocateFromRingBuffer(&g_d3d.resource_upload_ring_buffer, dst_byte_size, D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT);
-		D3D12_PLACED_SUBRESOURCE_FOOTPRINT placed_footprint = {};
-		placed_footprint.Offset = ring_buf_alloc.byte_offset;
-		placed_footprint.Footprint = pitch_footprint.Footprint;
+			RingBufferAllocation alloc = AllocateFromRingBuffer(&g_d3d.resource_upload_ring_buffer, dst_byte_size, D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT);
 
-		uint8_t* src_ptr = (uint8_t*)data_ptr;
-		uint8_t* dst_ptr = ring_buf_alloc.ptr;
+			// NOTE(daniel): For now I am carrying forward the assumption that rows are tightly patched for the src data.
+			// Hopefully, that's okay.
 
-		size_t dst_pitch = pitch_footprint.Footprint.RowPitch;
+			uint8_t *src_at = (uint8_t *)mips[mip_index];
+			uint8_t *dst_at = alloc.ptr;
 
-		for (size_t y = 0; y < row_count; ++y)
-		{
-			memcpy(dst_ptr, src_ptr, row_pitch);
+			size_t src_byte_size     = mip_width*mip_height*bytes_per_pixel;
+			size_t src_row_byte_size = src_byte_size / dst_row_count;
```

Comment:
This logic should be ok for all kinds of textures but it is a bit weird. It’s required because the DXT formats don’t fit into rows the way you might think they would. A 64x64 texture has 16 rows of 256 pixels each.